packages/form-state-manager/src/utils/state-manager-reducer.js

```diff
@@ -0,0 +1,69 @@
+export const REGISTER_FIELD = 'REGISTER_FIELD';
+export const UNREGISTER_FIELD = 'UNREGISTER_FIELD';
+
+const registerField = (state, { name, getFieldState }) => ({
+  ...state,
+  registeredFields: [...state.registeredFields, name],
+  fieldListeners: {
+    ...state.fieldListeners,
+    [name]: {
+      /**
+       * There can be multiple listeners checking the same field
+       * We need to check if we can destroy all listeners before we remove all of them
+       */
+      count: state.fieldListeners?.[name]?.count + 1 || 1,
+      getFieldState
+    }
+  }
+});
+
+const checkLastSubscrition = (fieldListeners, name) => fieldListeners?.[name]?.count === 1;
+
+const unregisterField = (state, { name, persistOnUnmount, value }) => {
+  const isLast = checkLastSubscrition(state.fieldListeners, name);
+  if (!isLast) {
+    return {
+      ...state,
+      fieldListeners: {
+        ...state.fieldListeners,
+        [name]: {
+          ...state.fieldListeners[name],
+          count: state.fieldListeners[name].count - 1
+        }
+      }
+    };
```

Comment:
just nitpick: I think this could be extracted to a separate function for better readability.